### PR TITLE
[STACK-2192]: Support for cascade deletion in vthunder flows

### PR DESCRIPTION
**Pull request by mohdadeebkhan**
## Description
Added support for cascade deletion of loadbalancer.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2192
https://a10networks.atlassian.net/browse/STACK-2181
https://a10networks.atlassian.net/browse/STACK-2190

## Technical Approach
1.On the basis of the cascade flag, deleted the loadbalancer's object.
If the cascade flag is true then delete the loadbalancer associated health monitors --> members --> pools --> l7policys --> listener --> loadbalancer.
2. Due to the overwriting of __heath_mon.type_ in _health_monitor_tasks.py_ task, health monitor unit test case were faling, so update the _health_monitor_tasks.py_ task

## Test Cases
1. Updated the delete loadbalancer method call arguments in _test_a10_load_balancer_flows.py_

## Manual Testing
1. Create the loadbalancer, listener with different protocol type, pools , members , healthmonitor, l7 policy, l7rule.
```
openstack loadbalancer create --vip-subnet-id vip_subnet --name lb1
openstack loadbalancer listener create --protocol TCP --protocol-port 8080 lb1 --name listener1
openstack loadbalancer listener create --protocol UDP --protocol-port 8081 lb1 --name listener2
openstack loadbalancer listener create --protocol HTTP --protocol-port 8083 lb1 --name listener3
openstack loadbalancer listener create --protocol TERMINATED_HTTPS --default-tls-container-ref http://10.64.3.129/key-manager/v1/containers/880ae4d1-c6d2-4bb9-9741-227be5a02a15 --protocol-port 9001 lb1 --name listener4
openstack loadbalancer pool create --protocol TCP --lb-algorithm ROUND_ROBIN --listener listener1 --name pool1
openstack loadbalancer pool create --protocol UDP --lb-algorithm ROUND_ROBIN --listener listener2 --name pool2
openstack loadbalancer pool create --protocol HTTP --lb-algorithm ROUND_ROBIN --listener listener3 --name pool3
openstack loadbalancer pool create --protocol HTTP --lb-algorithm ROUND_ROBIN --listener listener4 --name pool4
openstack loadbalancer member create --address 192.168.8.68  --subnet-id vip_subnet --protocol-port 80 --name member1 pool1
openstack loadbalancer member create --address 192.168.8.67  --subnet-id vip_subnet --protocol-port 80 --name member2 pool1
openstack loadbalancer member create --address 192.168.8.69  --subnet-id vip_subnet --protocol-port 81 --name member3 pool2
openstack loadbalancer member create --address 192.168.8.70  --subnet-id vip_subnet --protocol-port 82 --name member4 pool2
openstack loadbalancer member create --address 192.168.8.71  --subnet-id vip_subnet --protocol-port 83 --name member5 pool3
openstack loadbalancer member create --address 192.168.8.73  --subnet-id vip_subnet --protocol-port 84 --name member6 pool3
openstack loadbalancer healthmonitor create --delay 4 --timeout 3 --max-retries 3 --type HTTPS pool1
openstack loadbalancer healthmonitor create --delay 4 --timeout 3 --max-retries 3 --type HTTPS pool3
openstack loadbalancer l7policy create --action REJECT listener3 --name l7policy1
openstack loadbalancer l7policy create --action REJECT listener4 --name l7policy2
openstack loadbalancer l7rule create --type HEADER --compare-type EQUAL_TO --key abc --value xyz l7policy1

```
2. Result on vthnder after creation
```
vThunder(NOLICENSE)#show running-config
!Current configuration: 320 bytes
!Configuration last updated at 16:05:22 GMT Thu Mar 25 2021
!Configuration last saved at 16:05:25 GMT Thu Mar 25 2021
!64-bit Advanced Core OS (ACOS) version 4.1.4-GR1-P5, build 81 (Sep-08-2020,09:32)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
  name DataPort
  enable
  ip address dhcp
!
!
health monitor eb7ce1f1-de84-4770-b550-e49b1ffb5a65
  override-port 8080
  interval 4 timeout 3
  method https port 8080 expect response-code 200 url GET /
!
health monitor 54ce8144-84ed-4eed-b25b-738f7dc9a9c1
  override-port 8083
  interval 4 timeout 3
  method https port 8083 expect response-code 200 url GET /
!
slb server 85eaa_192_168_8_67 192.168.8.67
  port 80 tcp
!
slb server 85eaa_192_168_8_68 192.168.8.68
  port 80 tcp
!
slb server 85eaa_192_168_8_69 192.168.8.69
  port 81 udp
!
slb server 85eaa_192_168_8_70 192.168.8.70
  port 82 udp
!
slb server 85eaa_192_168_8_71 192.168.8.71
  port 83 tcp
!
slb server 85eaa_192_168_8_73 192.168.8.73
  port 84 tcp
!
slb service-group 5456ec65-d10e-4300-8e29-afaf5c8f2c2e tcp
  health-check eb7ce1f1-de84-4770-b550-e49b1ffb5a65
  member 85eaa_192_168_8_67 80
  member 85eaa_192_168_8_68 80
!
slb service-group 7fccad71-9667-431a-9f9a-b7946380eeec tcp
  health-check 54ce8144-84ed-4eed-b25b-738f7dc9a9c1
  member 85eaa_192_168_8_71 83
  member 85eaa_192_168_8_73 84
!
slb service-group b8425942-904c-400e-85a3-4dd7947ad4dc tcp
!
slb service-group cbe684f9-6eb6-4dbf-8e36-9ca0e4733f72 udp
  member 85eaa_192_168_8_69 81
  member 85eaa_192_168_8_70 82
!
slb template client-ssl 4e7fcf4a-471c-4493-b94d-0c86baf795d8
  cert mycert
  key mykey
!
slb virtual-server b8ba7650-c68e-4ece-bd99-7b9e5cd48947 192.168.8.159
  port 8080 tcp
    name 1996ff78-f135-4d94-8b48-b10b539f4777
    extended-stats
    service-group 5456ec65-d10e-4300-8e29-afaf5c8f2c2e
  port 8081 udp
    name f3fd47ed-f5d4-4709-83ec-05ca9bb6e2e1
    extended-stats
    service-group cbe684f9-6eb6-4dbf-8e36-9ca0e4733f72
  port 8083 http
    name a8293b00-3bb3-4006-a079-55f0a23caeb0
    extended-stats
    aflex b03b0d41-8914-4bb9-a698-cba9c341c343
    service-group 7fccad71-9667-431a-9f9a-b7946380eeec
  port 9001 https
    name 4e7fcf4a-471c-4493-b94d-0c86baf795d8
    extended-stats
    aflex aa306c03-fdcd-461d-8e6f-5cc8004e0749
    service-group b8425942-904c-400e-85a3-4dd7947ad4dc
    template client-ssl 4e7fcf4a-471c-4493-b94d-0c86baf795d8
!
slb virtual-server e594fdfb-1837-4ddb-8431-e8a599e7918e 192.168.8.123
!
!
cloud-services meta-data
  enable
  provider openstack
!
end
!Current config commit point for partition 0 is 0 & config mode is classical-mode
```
3. Delete the lb without cascade flag
```
stack@adeeb:~$ openstack loadbalancer delete lb1
Validation failure: Cannot delete Load Balancer bdd7b63d-5c46-4576-a289-83880469b642 - it has children (HTTP 400) (Request-ID: req-342caa0c-fa4f-4a9a-b7c0-8d34969452b7)
stack@adeeb:~$
```
4. Delete the lb with cascade flag
openstack loadbalancer delete lb1 --cascade
```
Mar 25 16:06:50 adeeb a10-octavia-worker[17704]: INFO a10_octavia.controller.queue.endpoint [-] Deleting load balancer 'b8ba7650-c68e-4ece-bd99-7b9e5cd48947'...
Mar 25 16:06:51 adeeb a10-octavia-worker[17704]: WARNING a10_octavia.controller.worker.tasks.nat_pool_tasks [-] Cannot delete Nat-pool(s) in flavor None as they are in use by another loadbalancer(s)
Mar 25 16:06:54 adeeb a10-octavia-worker[17704]: INFO barbicanclient.base [-] Calculated Containers uuid ref: containers/880ae4d1-c6d2-4bb9-9741-227be5a02a15
Mar 25 16:06:54 adeeb a10-octavia-worker[17704]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/71a0b9ac-866a-41e0-b4ab-c0a7a7787425
Mar 25 16:06:54 adeeb a10-octavia-worker[17704]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/4948fa8b-9df5-4bd2-bede-fe7ad9b5753c
Mar 25 16:06:54 adeeb a10-octavia-worker[17704]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/71a0b9ac-866a-41e0-b4ab-c0a7a7787425
Mar 25 16:06:54 adeeb a10-octavia-worker[17704]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/4948fa8b-9df5-4bd2-bede-fe7ad9b5753c
Mar 25 16:06:56 adeeb a10-octavia-worker[17704]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Attempting to connect vThunder device for connection.
Mar 25 16:06:56 adeeb a10-octavia-worker[17704]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 192.168.0.198:None

```
5. Result on vthunder after cascade deletion
```
vThunder(NOLICENSE)#show running-config
!Current configuration: 170 bytes
!Configuration last updated at 16:06:56 GMT Thu Mar 25 2021
!Configuration last saved at 16:07:04 GMT Thu Mar 25 2021
!64-bit Advanced Core OS (ACOS) version 4.1.4-GR1-P5, build 81 (Sep-08-2020,09:32)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
  name DataPort
  enable
  ip address dhcp
!
!
slb virtual-server e594fdfb-1837-4ddb-8431-e8a599e7918e 192.168.8.123
!
!
cloud-services meta-data
  enable
  provider openstack
!
end
!Current config commit point for partition 0 is 0 & config mode is classical-mode
```

**On Hardware device:**
1. Create the loadbalancer, listener with different protocol type, pools , members , healthmonitor, l7 policy, l7rule.[ as above]

2. Result of device
```
vThunder(NOLICENSE)#show running-config
!Current configuration: 337 bytes
!Configuration last updated at 08:58:26 GMT Fri Mar 26 2021
!Configuration last saved at 08:58:28 GMT Fri Mar 26 2021
!64-bit Advanced Core OS (ACOS) version 4.1.4-GR1-P5, build 81 (Sep-08-2020,09:32)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
!
interface ethernet 2
!
interface ethernet 3
!
interface ethernet 4
!
vrrp-a vrid 0
  floating-ip 192.168.8.120
!
health monitor 2320ff35-4b08-4f94-9755-d4a48a62c029
  override-port 8080
  interval 4 timeout 3
  method https port 8080 expect response-code 200 url GET /
!
health monitor 837a99bf-8b17-4e1e-8990-2e652fe15881
  override-port 8083
  interval 4 timeout 3
  method https port 8083 expect response-code 200 url GET /
!
slb server 85eaa_192_168_8_67 192.168.8.67
  port 80 tcp
!
slb server 85eaa_192_168_8_68 192.168.8.68
  port 80 tcp
!
slb server 85eaa_192_168_8_69 192.168.8.69
  port 81 udp
!
slb server 85eaa_192_168_8_70 192.168.8.70
  port 82 udp
!
slb server 85eaa_192_168_8_71 192.168.8.71
  port 83 tcp
!
slb server 85eaa_192_168_8_73 192.168.8.73
  port 84 tcp
!
slb service-group 4b2a1854-a52b-4110-8b34-5519003e745d tcp
!
slb service-group d4d5829d-236b-4a85-9196-2a2787de28b3 tcp
  health-check 2320ff35-4b08-4f94-9755-d4a48a62c029
  member 85eaa_192_168_8_67 80
  member 85eaa_192_168_8_68 80
!
slb service-group ff2a02d6-092a-4ce8-b795-cf612903a5a0 udp
  member 85eaa_192_168_8_69 81
  member 85eaa_192_168_8_70 82
!
slb service-group ffa0876a-4b43-4886-95f2-b19dd204af95 tcp
  health-check 837a99bf-8b17-4e1e-8990-2e652fe15881
  member 85eaa_192_168_8_71 83
  member 85eaa_192_168_8_73 84
!
slb template client-ssl bcc499d5-3cd3-4776-9a25-af14595ae6ec
  cert mycert
  key mykey
!
slb virtual-server 868a3e46-562a-49d5-bbb9-d5fc38e4f83e 192.168.8.169
  port 8080 tcp
    name 67ff9a42-69a7-4f30-9b38-374cd827ddf0
    extended-stats
    service-group d4d5829d-236b-4a85-9196-2a2787de28b3
  port 8081 udp
    name 80c5d656-1f4a-4293-8dfa-6c06c24382e5
    extended-stats
    service-group ff2a02d6-092a-4ce8-b795-cf612903a5a0
  port 8083 http
    name c0939495-28ad-4aea-a7fb-deb22a13baa6
    extended-stats
    aflex 128c983b-81d4-4c28-b54c-b9f0329da386
    service-group ffa0876a-4b43-4886-95f2-b19dd204af95
  port 9001 https
    name bcc499d5-3cd3-4776-9a25-af14595ae6ec
    extended-stats
    aflex d309215a-c386-47d9-90be-19a17db98ffa
    service-group 4b2a1854-a52b-4110-8b34-5519003e745d
    template client-ssl bcc499d5-3cd3-4776-9a25-af14595ae6ec
!
!
cloud-services meta-data
  enable
  provider openstack
!
end
!Current config commit point for partition 0 is 0 & config mode is classical-mode
vThunder(NOLICENSE)#
```
3. Delete the lb without cascade flag
```
stack@adeeb:~$ openstack loadbalancer list
+--------------------------------------+------+----------------------------------+---------------+---------------------+----------+
| id                                   | name | project_id                       | vip_address   | provisioning_status | provider |
+--------------------------------------+------+----------------------------------+---------------+---------------------+----------+
| 868a3e46-562a-49d5-bbb9-d5fc38e4f83e | lb1  | 85eaa5537bb74138976aa37c58d3bf64 | 192.168.8.169 | ACTIVE              | a10      |
+--------------------------------------+------+----------------------------------+---------------+---------------------+----------+
stack@adeeb:~$ openstack loadbalancer  delete  lb1
Validation failure: Cannot delete Load Balancer 868a3e46-562a-49d5-bbb9-d5fc38e4f83e - it has children (HTTP 400) (Request-ID: req-ec9048da-8670-4aea-a7ab-fc019d921454)
```
4. Delete the lb with cascade flag
openstack loadbalancer delete lb1 --cascade
```
Mar 26 09:20:18 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.queue.endpoint [-] Deleting load balancer '868a3e46-562a-49d5-bbb9-d5fc38e4f83e'...
Mar 26 09:20:21 adeeb a10-octavia-worker[1121]: INFO barbicanclient.base [-] Calculated Containers uuid ref: containers/880ae4d1-c6d2-4bb9-9741-227be5a02a15
Mar 26 09:20:21 adeeb a10-octavia-worker[1121]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/71a0b9ac-866a-41e0-b4ab-c0a7a7787425
Mar 26 09:20:21 adeeb a10-octavia-worker[1121]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/4948fa8b-9df5-4bd2-bede-fe7ad9b5753c
Mar 26 09:20:21 adeeb a10-octavia-worker[1121]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/71a0b9ac-866a-41e0-b4ab-c0a7a7787425
Mar 26 09:20:21 adeeb a10-octavia-worker[1121]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/4948fa8b-9df5-4bd2-bede-fe7ad9b5753c
Mar 26 09:20:22 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.a10_network_tasks [-] VRID floating IP: 192.168.8.120 deleted
Mar 26 09:20:22 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Attempting to connect vThunder device for connection.
Mar 26 09:20:22 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 10.0.0.72:shared
```
5. Result on device after deletion
```
vThunder(NOLICENSE)#show running-config
!Current configuration: 187 bytes
!Configuration last updated at 09:20:22 GMT Fri Mar 26 2021
!Configuration last saved at 09:20:29 GMT Fri Mar 26 2021
!64-bit Advanced Core OS (ACOS) version 4.1.4-GR1-P5, build 81 (Sep-08-2020,09:32)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
!
interface ethernet 2
!
interface ethernet 3
!
interface ethernet 4
!
!
!
cloud-services meta-data
  enable
  provider openstack
!
end
!Current config commit point for partition 0 is 0 & config mode is classical-mode
vThunder(NOLICENSE)#
```
6. Check for existing functionality
Create only loadbalancer[without children] then delete it.
```
openstack loadbalancer create --vip-subnet-id vip_subnet --name lb2
logs:
Mar 26 09:31:22 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.queue.endpoint [-] Creating load balancer '694377ee-fc75-486f-8896-65cc74b5061a'...
Mar 26 09:31:22 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.a10_database_tasks [-] Successfully created vthunder entry in database.
Mar 26 09:31:34 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Partition shared created
Mar 26 09:31:35 adeeb a10-octavia-worker[1121]: INFO octavia.controller.worker.tasks.database_tasks [-] Mark ACTIVE in DB for load balancer id: 694377ee-fc75-486f-8896-65cc74b5061a
Mar 26 09:31:35 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 10.0.0.72:shared

openstack loadbalancer  delete  lb2
logs:
Mar 26 09:31:50 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.queue.endpoint [-] Deleting load balancer '694377ee-fc75-486f-8896-65cc74b5061a'...
Mar 26 09:31:51 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.a10_network_tasks [-] VRID floating IP: 192.168.8.77 deleted
Mar 26 09:31:51 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Attempting to connect vThunder device for connection.
Mar 26 09:31:51 adeeb a10-octavia-worker[1121]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 10.0.0.72:shared
```

 **NATPOOL Deletion testing:**
 1) Create Flavor profile, flavor and loadbalancer
 ```
openstack loadbalancer flavorprofile create --name fp_snat --provider a10 --flavor-data '{"nat-pool":{"pool-name":"pool1", "start-address":"192.168.8.201", "end-address":"192.168.8.210", "netmask":"/24", "gateway":"192.168.8.1" }}' 
openstack loadbalancer flavor create --name f_snat --flavorprofile fp_snat --description "SNAT Flavor1" --enable
openstack loadbalancer create --flavor f_snat --vip-subnet-id vip_subnet --vip-address 192.168.8.57 --name lb1
```
2) Result on vthunder
```
vrrp-a vrid 0
  floating-ip 192.168.8.247
!
ip nat pool pool1 192.168.8.201 192.168.8.210 netmask /24 gateway 192.168.8.1
!
health monitor 09c03dba-937d-4b1f-8d3b-253551cbacec
  override-port 8080
  interval 4 timeout 3
  method https port 8080 expect response-code 200 url GET /
!
health monitor eeb4c555-4f55-4eeb-b11a-2b3012ff6f8e
  override-port 8083
  interval 4 timeout 3
  method https port 8083 expect response-code 200 url GET /
!
slb server 85eaa_192_168_8_67 192.168.8.67
  port 80 tcp
!
slb server 85eaa_192_168_8_68 192.168.8.68
  port 80 tcp
!
slb server 85eaa_192_168_8_69 192.168.8.69
  port 81 udp
!
slb server 85eaa_192_168_8_70 192.168.8.70
  port 82 udp
!
slb server 85eaa_192_168_8_71 192.168.8.71
  port 83 tcp
!
slb server 85eaa_192_168_8_73 192.168.8.73
  port 84 tcp
!
slb service-group 3fda9d18-fb61-4900-b742-430a0de85eb3 tcp
  health-check 09c03dba-937d-4b1f-8d3b-253551cbacec
  member 85eaa_192_168_8_67 80
  member 85eaa_192_168_8_68 80
!
slb service-group 5c64346f-2ebb-4eb2-b235-9fd79abcd25d tcp
!
slb service-group 6e30996a-3243-4da7-8648-15885f51b7b2 tcp
  health-check eeb4c555-4f55-4eeb-b11a-2b3012ff6f8e
  member 85eaa_192_168_8_71 83
  member 85eaa_192_168_8_73 84
!
slb service-group 972b0916-4431-47f7-84b9-4e5bea0df77e udp
  member 85eaa_192_168_8_69 81
  member 85eaa_192_168_8_70 82
!
slb template client-ssl cd4d3532-8492-40ab-9f38-a4b2f9d8ddf6
  cert mycert
  key mykey
!
slb virtual-server f5a3e2ae-87cb-423e-8043-26be8a04ec30 192.168.8.57
  port 8080 tcp
    name 42181f8e-5279-4378-85e5-0b6f391deee1
    extended-stats
    source-nat pool pool1
    service-group 3fda9d18-fb61-4900-b742-430a0de85eb3
  port 8081 udp
    name d0ba3e37-427e-4bbb-ad22-f57cfed0e8e6
    extended-stats
    source-nat pool pool1
    service-group 972b0916-4431-47f7-84b9-4e5bea0df77e
  port 8083 http
    name fd9ea5d5-b6ba-48a3-a722-8fcd421b9507
    extended-stats
    aflex 7ac4897b-a464-4652-9484-133c438c0ed2
    source-nat pool pool1
    service-group 6e30996a-3243-4da7-8648-15885f51b7b2
  port 9001 https
    name cd4d3532-8492-40ab-9f38-a4b2f9d8ddf6
    extended-stats
    aflex 21733e3c-b96d-4632-a79f-d7de8dc2ca4b
    source-nat pool pool1
    service-group 5c64346f-2ebb-4eb2-b235-9fd79abcd25d
    template client-ssl cd4d3532-8492-40ab-9f38-a4b2f9d8ddf6
!
```
 
3) Delete the loadbalaner with cascade option
openstack loadbalancer delete lb1 --cascade
```
Mar 31 08:11:40 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.queue.endpoint [-] Deleting load balancer 'f5a3e2ae-87cb-423e-8043-26be8a04ec30'...
Mar 31 08:11:45 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.a10_database_tasks [-] Successfully deleted nat pool entry in database.
Mar 31 08:11:45 adeeb a10-octavia-worker[19105]: INFO barbicanclient.base [-] Calculated Containers uuid ref: containers/880ae4d1-c6d2-4bb9-9741-227be5a02a15
Mar 31 08:11:46 adeeb a10-octavia-worker[19105]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/71a0b9ac-866a-41e0-b4ab-c0a7a7787425
Mar 31 08:11:46 adeeb a10-octavia-worker[19105]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/4948fa8b-9df5-4bd2-bede-fe7ad9b5753c
Mar 31 08:11:46 adeeb a10-octavia-worker[19105]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/71a0b9ac-866a-41e0-b4ab-c0a7a7787425
Mar 31 08:11:46 adeeb a10-octavia-worker[19105]: INFO barbicanclient.base [-] Calculated Secrets uuid ref: secrets/4948fa8b-9df5-4bd2-bede-fe7ad9b5753c
Mar 31 08:11:47 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.a10_network_tasks [-] VRID floating IP: 192.168.8.247 deleted
Mar 31 08:11:47 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Attempting to connect vThunder device for connection.
Mar 31 08:11:47 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 10.0.0.72:shared
```
4) Result after deletion
```
!64-bit Advanced Core OS (ACOS) version 4.1.4-GR1-P5, build 81 (Sep-08-2020,09:32)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
!
interface ethernet 2
!
interface ethernet 3
!
interface ethernet 4
!
!
!
cloud-services meta-data
  enable
  provider openstack
!
end
```

Natpool list test results:
1) create flavorprofile, flavor and loadbalancer with natpool list
```
openstack loadbalancer flavorprofile create --name fp_snat_list --provider a10 --flavor-data '{"nat-pool":{"pool-name":"pool1", "start-address":"192.168.8.101", "end-address":"192.168.8.102", "netmask":"/24", "gateway":"192.168.8.1"}, "nat-pool-list":[{"pool-name":"pool2", "start-address":"192.168.8.111", "end-address":"192.168.8.112", "netmask":"/24", "gateway":"192.168.8.1"}, {"pool-name":"pool3", "start-address":"192.168.8.120", "end-address":"192.168.8.121", "netmask":"/24", "gateway":"192.168.8.1"}]}'
openstack loadbalancer flavor create --name f_snat_list --flavorprofile fp_snat_list --description "flaovr all test1" --enable
openstack loadbalancer create --flavor f_snat_list --vip-subnet-id vip_subnet --vip-address 192.168.8.59 --name lb1
openstack loadbalancer listener create --protocol HTTP --protocol-port 8083 lb1 --name  vport3
openstack loadbalancer pool create --protocol HTTP --lb-algorithm ROUND_ROBIN --listener vport3 --name sg1
``` 
2) Result on vthunder
```
vrrp-a vrid 0
  floating-ip 192.168.8.43
!
ip nat pool pool1 192.168.8.101 192.168.8.102 netmask /24 gateway 192.168.8.1
!
ip nat pool pool2 192.168.8.111 192.168.8.112 netmask /24 gateway 192.168.8.1
!
ip nat pool pool3 192.168.8.120 192.168.8.121 netmask /24 gateway 192.168.8.1
!
slb server 85eaa_192_168_8_132 192.168.8.132
  port 80 tcp
!
slb service-group aa894ade-a98f-4d21-b179-151136915d8d tcp
  member 85eaa_192_168_8_132 80
!
slb virtual-server 1d1c6aad-7c2a-459e-a300-f544535dd33f 192.168.8.59
  port 8083 http
    name 3fe5eeeb-bb01-4317-bd33-ebf51435f79e
    extended-stats
    source-nat pool pool1
    service-group aa894ade-a98f-4d21-b179-151136915d8d
!
```
3) Delete the lb
openstack loadbalancer delete lb1 --cascade
```
Mar 31 09:05:11 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.queue.endpoint [-] Deleting load balancer '1d1c6aad-7c2a-459e-a300-f544535dd33f'...
Mar 31 09:05:12 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.a10_database_tasks [-] Successfully deleted nat pool entry in database.
Mar 31 09:05:13 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.a10_network_tasks [-] VRID floating IP: 192.168.8.43 deleted
Mar 31 09:05:13 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Attempting to connect vThunder device for connection.
Mar 31 09:05:13 adeeb a10-octavia-worker[19105]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 10.0.0.72:shared
```
4) Result on vthunder after deletion
```
!64-bit Advanced Core OS (ACOS) version 4.1.4-GR1-P5, build 81 (Sep-08-2020,09:32)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
!
interface ethernet 2
!
interface ethernet 3
!
interface ethernet 4
!
!
!
cloud-services meta-data
  enable
  provider openstack
!
```

Natpoot is not deleted when there more than one lb
1)Create two lbs.
2) Result on vthunder
```
!
vrrp-a vrid 0
  floating-ip 192.168.8.99
!
ip nat pool pool1 192.168.8.201 192.168.8.210 netmask /24 gateway 192.168.8.1
!
slb virtual-server 5197cb5e-24d8-4bd2-9396-c46d84a3c2fc 192.168.8.58
!
slb virtual-server b83724c9-5e42-418d-8255-13f218ecbd57 192.168.8.57
!
```
3) Delete one lb
logs:
```
Mar 31 07:24:12 adeeb a10-octavia-worker[8903]: INFO a10_octavia.controller.queue.endpoint [-] Deleting load balancer '5197cb5e-24d8-4bd2-9396-c46d84a3c2fc'...
Mar 31 07:24:13 adeeb a10-octavia-worker[8903]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Attempting to connect vThunder device for connection.
Mar 31 07:24:13 adeeb a10-octavia-worker[8903]: WARNING a10_octavia.controller.worker.tasks.nat_pool_tasks [-] Cannot delete Nat-pool(s) in flavor 4da82b5c-21bf-4876-8666-929bbef82833 as they are in use by another loadbalancer(s)
Mar 31 07:24:13 adeeb a10-octavia-worker[8903]: INFO a10_octavia.controller.worker.tasks.vthunder_tasks [-] Performing write memory for thunder - 10.0.0.72:shared
```
4) Result on vthunder after deletion
```
!
vrrp-a vrid 0
  floating-ip 192.168.8.99
!
ip nat pool pool1 192.168.8.201 192.168.8.210 netmask /24 gateway 192.168.8.1
!
slb virtual-server b83724c9-5e42-418d-8255-13f218ecbd57 192.168.8.57
!
!
```

### CODE REVIEW

**Comment by ytsai-a10** [comment]
- Please help consist the naming rules for new APIs. for example
  - _cascade_ for cascade flow/subflow
  - _internal_ for internal flow/subflows for _cascade_ flows
  - consist with octavia naming rule, like 'get_cascade_delete_...'  or '..._internal_flow/subflow'
- I think we need more test cases for this feature. please list your test cases
  - different listener protocol type
  - test without using cascade

**Review by ytsai-a10 — DISMISSED** [review]

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]
This adds a lot of LOC to our codebase. There isn't a real need for that either. For example there is no difference between  `get_cascade_delete_ssl_certificate_flow` and 
`get_ssl_certificate_delete_flow` outside of the addition of misleading names. I believe the same can be said for the members.

Even for the pool delete flow, we can just delete the members and health monitor first and then delete the pool after. This all happens in linear flows so we can be certain that those objects are deleted first.

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py [hunk 1]
```diff
@@ -505,3 +517,59 @@ def get_delete_lb_vrid_subflow(self):
...
+                 the listeners stored properly
+        """
+        listeners_delete_flow = unordered_flow.Flow('listener_delete_flow')
+        store = {}
+        compute_flag = lb.amphorae
+        for listener in lb.listeners:
+            for l7policy in listener.l7policies:
+                l7policy_name = 'l7policy_' + l7policy.id
+                store[l7policy_name] = l7policy
+                listeners_delete_flow.add(
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:536]: This will lead to an unintended side effect where an attempt is made to delete an L7 policy before the listener as it's being added to an unordered flow.

Instead, create an UnorderedFlow to start. On each iteration, create a linear flow and add the listener delete flow to it then add the L7 policy delete flow to it.

For 3 listeners this would look like

```
UnorderedFlow( LinearFlow( ListenerDeleteFlow() -> L7PolicyDeleteFlow() ),
               LinearFlow( ListenerDeleteFlow() -> L7PolicyDeleteFlow() )
               LinearFlow( ListenerDeleteFlow() -> L7PolicyDeleteFlow() )
````


This model can be extended to the member , health monitor, pool side of things too.

Technically this entire construct could be handled in one function. No real need to separate Listener from pool. My guess was that it was done to increase read ability.

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]

**Comment by ytsai-a10** [comment]
Please ask JiYen help you run automation before you merge in to master

File: a10_octavia/controller/worker/tasks/health_monitor_tasks.py [hunk 2]
```diff
@@ -57,17 +57,18 @@ def execute(self, listeners, health_mon, vthunder, flavor=None):
                 args.update({'monitor': flavors})
 
         try:
-            health_mon.type = openstack_mappings.hm_type(self.axapi_client, health_mon.type)
+            health_mon_type = openstack_mappings.hm_type(self.axapi_client, health_mon.type)
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/tasks/health_monitor_tasks.py:60]: Please revert this change. The root cause has been fixed in another PR

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/controller/worker/tasks/health_monitor_tasks.py (hunk 2 above)
Comment (mohdadeebkhan) [reply a10_octavia/controller/worker/tasks/health_monitor_tasks.py]: Reverted

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py (hunk 1 above)
Comment (mohdadeebkhan) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py]: Done the changes as discussed over teams

**Review by ytsai-a10 — APPROVED** [review]

**Review by hthompson-a10 — APPROVED** [review]

**Review by JiYen-Chen — APPROVED** [review]

**Review by JiYen-Chen — APPROVED** [review]
Automation was done and no important was found.
